pkg/virt-launcher/virtwrap/network/network.go

```diff
@@ -0,0 +1,250 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/vishvananda/netlink"
+
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/precond"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/network/dhcp"
+
+	lmf "github.com/subgraph/libmacouflage"
+)
+
+const (
+	podInterface     = "eth0"
+	macVlanIfaceName = "macvlan0"
+	macVlanFakeIP    = "10.11.12.13/24"
+	guestDNS         = "8.8.8.8"
+)
+
+type VIF struct {
+	Name    string
+	IP      netlink.Addr
+	MAC     net.HardwareAddr
+	Gateway net.IP
+}
+
+// SetupDefaultPodNetwork will prepare the pod management network to be used by a virtual machine
+// which will own the pod network IP and MAC. Pods MAC address will be changed to a
+// random address and IP will be deleted. This will also create a macvlan device with a fake IP.
+// DHCP server will be started and bounded to the macvlan interface to server the original pod ip
+// to the guest OS
+func SetupDefaultPodNetwork(domain *api.Domain) error {
+	precond.MustNotBeNil(domain)
+	// Get IP and MAC
+	// Change eth0 MAC
+	// Create macvlan and set fake address
+	// remove eth0 IP
+	// Start DHCP
+
+	nic := &VIF{Name: podInterface}
+	link, err := netlink.LinkByName(podInterface)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to get a link for interface: %s", podInterface)
+		return err
+	}
+
+	// get IP address
+	addrList, err := netlink.AddrList(link, netlink.FAMILY_V4)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to get an ip address for %s", podInterface)
+		return err
+	}
+	if len(addrList) == 0 {
+		return fmt.Errorf("No IP address found on %s", podInterface)
+	}
+	nic.IP = addrList[0]
+
+	// Get interface gateway
+	routes, err := netlink.RouteList(link, netlink.FAMILY_V4)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to get routes for %s", podInterface)
+		return err
+	}
+	if len(routes) == 0 {
+		return fmt.Errorf("No gateway address found in routes for %s", podInterface)
+	}
+	nic.Gateway = routes[0].Gw
+
+	// Get interface MAC address
+	mac, err := GetMacDetails(podInterface)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to get MAC for %s", podInterface)
+		return err
+	}
+	nic.MAC = mac
+
+	// Remove IP from POD interface
+	err = netlink.AddrDel(link, &nic.IP)
+
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to delete link for interface: %s", podInterface)
+		return err
+	}
+
+	// Set interface link to down to change its MAC address
+	err = netlink.LinkSetDown(link)
```

Comment:
Since we are changing the pod interface, we either first need to collect all data about the interface, and save it in an extra file for retries (that is what the code right now tries to do), or we say that a pod failed, if we can't set this up (exit with non-zero exit code).

Reply:
@vladikr is that addressed?